File: day04/part2.hs
import Data.List
import qualified Data.Map as M
import System.Environment
import System.FilePath

type Block = [String]

type Entry = (String, String)

type Passport = M.Map String String

checkRange :: Int -> Int -> Int -> Bool
checkRange lo hi n = lo <= n && n <= hi

hexchars :: String
hexchars = "0123456789abcdef"

eyecolors :: [String]
eyecolors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]

height :: String -> (String, Int)
height s = (reverse revUnit, read (reverse revN))
  where
    (revUnit, revN) = splitAt 2 $ reverse s

containedIn :: Eq a => [a] -> a -> Bool
containedIn l x = x `elem` l

rules :: M.Map String (String -> Bool)
rules =
  M.fromList
    [ ("byr", (checkRange 1920 2002 .) $ read),
      ("iyr", (checkRange 2010 2020 .) $ read),
      ("eyr", (checkRange 2020 2030 .) $ read),
      ( "hgt",
        \s -> case height s of
          ("cm", n) -> checkRange 150 193 n
          ("in", n) -> checkRange 59 76 n
          _ -> False
      ),
      ("hcl", \s -> ("#" == take 1 s) && all (containedIn hexchars) (drop 1 s)),
      ("ecl", containedIn eyecolors),
      ("pid", \s -> length s == 9)
    ]

split :: Eq a => a -> [a] -> [[a]]
split d s = case span (/= d) s of
  (x, _ : ts) -> x : split d ts
  (x, []) -> x : []

-- https://stackoverflow.com/a/23979003
sameElems :: Eq a => [a] -> [a] -> Bool
sameElems xs ys = null (xs \\ ys) && null (ys \\ xs)

entry :: String -> Entry
entry s = case split ':' s of
  [a, b] -> (a, b)

blocks :: [String] -> [Block]
blocks [] = []
blocks ls = a' : blocks b
  where
    (a, b) = case span (/= "") ls of
      ([], b) -> ([], tail b)
      (a, b) -> (a, b)
    a' = concatMap (split ' ') a

passport :: Block -> Passport
passport xs = M.fromList $ map entry xs

containsRequired :: Passport -> Bool
containsRequired p = intersect (map fst $ M.toList p) (M.keys rules) `sameElems` (M.keys rules)

checkPassportEntry :: Entry -> Bool
checkPassportEntry (k, v) = case M.lookup k rules of
  Just r -> r v
  Nothing -> True

checkPassport :: Passport -> Bool
checkPassport p = containsRequired p && all checkPassportEntry (M.toList p)

compute :: [String] -> Int
compute ls = length . (filter checkPassport) . (map passport) . blocks $ ls

main :: IO ()
main = do
  exePath <- getExecutablePath
  content <- readFile (replaceFileName exePath "input.txt")
  putStrLn . show . compute . lines $ content

File: day04/part2.py
_TESTCASE = """\
eyr:1972 cid:100
hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926

iyr:2019
hcl:#602927 eyr:1967 hgt:170cm
ecl:grn pid:012533040 byr:1946

hcl:dab227 iyr:2012
ecl:brn hgt:182cm pid:021572410 eyr:2020 byr:1992 cid:277

hgt:59cm ecl:zzz
eyr:2038 hcl:74454a iyr:2023
pid:3556412378 byr:2007

pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980
hcl:#623a2f

eyr:2029 ecl:blu cid:129 byr:1989
iyr:2014 pid:896056539 hcl:#a97842 hgt:165cm

hcl:#888785
hgt:164cm byr:2001 iyr:2015 cid:88
pid:545766238 ecl:hzl
eyr:2022

iyr:2010 hgt:158cm hcl:#b6652a ecl:blu byr:1944 eyr:2021 pid:093154719
""".strip().split(
    "\n\n"
)


# noinspection PyUnboundLocalVariable
def compute(data):
    """
    >>> compute(_TESTCASE)
    4
    """

    def range_checker(lo, hi):
        return lambda x: x.isnumeric() and lo <= int(x) <= hi

    valid_per_unit_hgt_ranges = {"cm": range(150, 194), "in": range(59, 77)}
    valid_hexdigits = frozenset("0123456789abcdef")
    valid_eyecolors = frozenset(("amb", "blu", "brn", "gry", "grn", "hzl", "oth"))

    rules = {
        "byr": range_checker(1920, 2002),
        "iyr": range_checker(2010, 2020),
        "eyr": range_checker(2020, 2030),
        "hgt": lambda x: (
            (unit := x[-2:]) in valid_per_unit_hgt_ranges
            and (num_s := x[:-2]).isnumeric()
            and int(num_s) in valid_per_unit_hgt_ranges[unit]
        ),
        "hcl": lambda x: (
            x.startswith("#") and len(hexnum_s := x[1:]) == 6 and set(hexnum_s) <= valid_hexdigits
        ),
        "ecl": lambda x: x in valid_eyecolors,
        "pid": lambda x: x.isnumeric() and len(x) == 9,
    }
    required = frozenset(rules)
    passports = (
        dict(field.split(":") for line in block.splitlines() for field in line.split(" "))
        for block in data
    )

    return sum(
        1
        for passport in passports
        if required <= passport.keys()
        if all(rules[key](val) for key, val in passport.items() if key in rules)
    )


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")

    with input_path.open() as f:
        print(compute(f.read().strip().split("\n\n")))


if __name__ == "__main__":
    main()
